### DISCUSSION TITLE
Should I be able to quit Zim via osascript?

### DISCUSSION THREAD

**Discussion by reagle** [zim-desktop-wiki/zim-desktop-wiki#1884] [Q&A]
```
╰─➤  osascript -e 'quit app "Zim"'                                                            1 ↵
0:14: execution error: Zim got an error: Can’t continue quit. (-1708)
```

**Comment by jaap-karssenberg** [comment]
What is "osascript" ?

Op di 25 jan. 2022 19:09 schreef Joseph Reagle ***@***.***>:

**Reply by dehesselle** [reply]
It's the scripting language (for automation) on macOS. Kind of like autohotkey.

**Comment by dehesselle** [comment]
Without having looked that one up, based on my experience:
The answer is probably "no" at the moment. I guess that this "quit" comes in the form of an Apple event which Zim (or GTK) needs to act upon. Right now, that event is not handled in the code. (Same goes for "open a notebook by double clicking it in Finder".)

If someone develops a "Hello World"-style Python code to demonstrate how to do that, I'm all ears.

**Reply by reagle** [reply]
Thank you. I don't know much about these things, so I've no idea. The reason I ask is that when I sync my files across my desktops I thought it best to ensure that Zim is **not** still open on the other computer. I send a `killall` via ssh but that's not as gentle as a `osascript quit,` which allows files to save files and cleanup normally.

**Reply by dehesselle** [reply]
Your idea is good, I think we just need another way to say "please quit" for now.
Have a look here: https://stackoverflow.com/questions/3690167/how-can-one-invoke-a-keyboard-shortcut-from-within-an-applescript
Apparently you can send keystrokes as well. Give it a try.

**Reply by introt** [reply]
> I send a killall via ssh but that's not as gentle as a osascript quit, which allows files to save files and cleanup normally.

Unless you specify a different signal, `killall` should send a `SIGTERM` which is handled gracefully:

https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/e300a32e5bfcee4fd9368c91e1098688137f4ce0/zim/main/__init__.py#L830-L839

Compare with the "Quit" menu item:

https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/e300a32e5bfcee4fd9368c91e1098688137f4ce0/zim/gui/uiactions.py#L244-L253

There are some mishandled signals (events, interrupts, ...) (see #871), but a plain-ol' `SIGTERM` should be safe.
